### ISSUE TITLE
ibm_db.connect different in python 2.7 compared to python 3.5

### ISSUE THREAD

**Issue by klaasbrant** [ibmdb/python-ibmdb#339]
When doing in python 2.7.12
**import ibm_db
db_conn = ibm_db.connect('dsn=sample', '', '')**
I get an error:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Exception: [IBM][CLI Driver] SQL1531N  The connection failed because the name specified with the DSN connection string keyword could not be found in either the db2dsdriver.cfg configuration file or the db2cli.ini configuration file.  Data source name specified in the connection string: "SAMPLE". SQLCODE=-1531

However if I do the same thing in python 3.5.2 on the same machine then it nicely connect to the local database sample.

Version of ibm_db is 2.0.9 in both versions
Local DB2 is V11.1.2.2 in ubuntu linux 16.04

**Comment by SabaKauser** [comment]
Hello Klaas,

The error implies that with python 2.7.12, the driver is referencing thin client driver i.e clidriver installation under your egg folder or it is referring to an existing installation that does not have catalogs and hence requires the connection details to be given in db2dsdriver.cfg configuration file or pass the full connection string with host and port name.
Where as with python 3.5 example, you seem to be executing using the full client (sqllib/lib64) and this library has access to catalogs and can read the connection information from there.

**Comment by SabaKauser** [comment]
Hello Klaas,
can you share ldd output of ibm_db.so and ibm_db.cpython-35m-x86_64-linux-gnu.so.
There seems to something different in the env. May be you need to check LD_LIBRARY_PATH to see that you are pointing to intended library of clidriver.
for e.g: if 3.5 is picking full client, the library path would be pointing to sqllib/lib* and with 2.7, it would be pointing to clidriver/lib.

**Comment by klaasbrant** [comment]
Hello Saba,
Another piece of the puzzle. The pip is done with sudo so runs as root to install for all users. root does not have the environment variables set as DB2 user. No problem for 3.5 but is a problem for 2.7
When installing with pip --user the 2.7 is no problem and fast (because it copies from SQLLIB?, no compiles!)

**Comment by SabaKauser** [comment]
Hello Klaas,
We do not copy from sqllib. But yes, if IBM_DB_HOME is set to sqllib, we would not download the clidriver(a.k.a thin client). And the python driver would be compiled referring to the libdb2.so as picked while building.
Try this from python 2.7:
export IBM_DB_HOME=<path to sqllib upto sqllib>
export IBM_DB_LIB=<path to lib64>

then pip install. You should see the install happening quickly and it should connect with just database name.

**Comment by klaasbrant** [comment]
Hi Saba,
I have tried a view scenarios and it comes done to: When db2profile set the environment variables then 2.7 and 3.5 work perfect. When not set (like in the case of sudo, root is NOT a DB2 user) then 2.7 goes for ODBC thin client and 3.5 goes for full client. So the setup.py work different in these two cases. Why?
Maybe just need an extra line in the doc saying that you have to make sure the the db2profile is executed before using pip.

**Comment by SabaKauser** [comment]
Hello Klaas,

does the db2profile contain IBM_DB_HOME setting or just the library path setting.

**Comment by SabaKauser** [comment]
Thanks Klaas,
This makes sense now.

I will add necessary information to README. May be on monday as I am on vacation.

**Comment by klaasbrant** [comment]
Hi Saba,
If the README explains that DB2 environment variables are expected to be set (e.g. by executing db2profile from sqllib) that would be great. Maybe a hint that if pip is used by sudo you want to use the -E flag to copy the environment to the sudo environment. Maybe a warning from setup if the environment variables are not set?
But... remains the original mystery. Why does setup act different in 2.7 and 3.5 when the environment variables are missing? And what is the correct behaviour?

**Comment by SabaKauser** [comment]
Hello Klaas,
python ibm_db's setup.py does not differentiate between python versions. It performs plain check to see if IBM_DB_HOME is set. If not set, it would download the driver automatically and use it while compiling to resolve symbols. 
When you execute applications via the driver, the shared library is that ibm_db.*so refers is picked by OS. If you have it set to pick full client or thin client, same would be used accordingly. 

In your case, with python 2.7 , we did not see IBM_DB_HOME set, hence we downloaded the driver and also. While executing, library path had precedence to thin client and this configuration needs the connection details to be passed in connection string or in db2dsdriver.cfg. 
However, with Python 3.5 command window, you had the variable set to full client location and while executing also, it used full client and read the catalogs.

I have both thin client and full client in my machine. I set the library path corresponding to which client I want to use post installation.
